python/demo/gsoc2022/demo_scattering_boundary_conditions.py

```diff
@@ -0,0 +1,451 @@
+    exit(0)
+# -
+
+# The following function is used for defining the background field.
+# The inputs to the function are the angle $\theta$, the background
+# refractive index $n_b$ and the vacuum wavevector $k_0$. The
+# function returns the expression $ \mathbf{E}_b = -\sin
+# \theta e^{j (k_xx+k_yy)}\hat{\mathbf{u}}_x
+# + \cos\theta e^{j (k_xx+k_yy)}\hat{\mathbf{u}}_y$.
+
+# +
+
+
+class background_electric_field:
+
+    def __init__(self, theta, n_b, k0):
+        self.theta = theta
+        self.k0 = k0
+        self.n_b = n_b
+
+    def eval(self, x):
+
+        kx = self.n_b * self.k0 * np.cos(self.theta)
+        ky = self.n_b * self.k0 * np.sin(self.theta)
+        phi = kx * x[0] + ky * x[1]
+
+        ax = np.sin(self.theta)
+        ay = np.cos(self.theta)
+
+        return (-ax * np.exp(1j * phi), ay * np.exp(1j * phi))
+# -
+
+# +
+# Definition of the radial distance from the center
+
+
+def radial_distance(x):
+    return np.sqrt(x[0]**2 + x[1]**2)
+
+# Definition of the curl for a 2d vector
+
+
+def curl_2d(a):
+
+    ay_x = a[1].dx(0)
+    ax_y = a[0].dx(1)
+
+    c = as_vector((0, 0, ay_x - ax_y))
+
+    return c
+
+
+# Constant definition
+um = 10**-6  # micron
+nm = 10**-9  # nanometer
+pi = np.pi
+epsilon_0 = 8.8541878128 * 10**-12
+mu_0 = 4 * pi * 10**-7
+
+# Radius of the wire and of the boundary of the domain
+radius_wire = 0.050 * um
+radius_dom = 1 * um
+
+# The smaller the mesh_factor, the finer is the mesh
+mesh_factor = 1.2
+
+# Finite element degree
+degree = 3
+
+# Wavelength sweep
+wl0 = 0.4 * um
+
+# Background refractive index
+n_bkg = 1.33
+eps_bkg = n_bkg**2
+
+# Mesh size inside the wire
+in_wire_size = mesh_factor * 7 * nm
+
+# Mesh size at the boundary of the wire
+on_wire_size = mesh_factor * 3 * nm
+
+# Mesh size in the vacuum
+bkg_size = mesh_factor * 60 * nm
+
+# Mesh size at the boundary
+boundary_size = mesh_factor * 30 * nm
+
+# Tags for the subdomains
+au_tag = 1          # gold wire
+bkg_tag = 2         # background
+boundary_tag = 3    # boundary
+
+model = generate_mesh_wire(
+    radius_wire, radius_dom, in_wire_size, on_wire_size, bkg_size,
+    boundary_size, au_tag, bkg_tag, boundary_tag)
+
+mesh, cell_tags, facet_tags = gmsh_model_to_mesh(
+    model, cell_data=True, facet_data=True, gdim=2)
+
+MPI.COMM_WORLD.barrier()
+
+# Definition of finite element for the electric field
+curl_el = ufl.FiniteElement("N1curl", mesh.ufl_cell(), 3)
+V = fem.FunctionSpace(mesh, curl_el)
+
+# Wavevector of the background field
+k0 = 2 * np.pi / wl0
+
+# Angle of incidence of the background field
+deg = np.pi / 180
+theta = 45 * deg
+
+# Plane wave function
+f = background_electric_field(theta, n_bkg, k0)
+Eb = fem.Function(V)
+Eb.interpolate(f.eval)
+
+# Function r = radial distance from the (0, 0) point
+lagr_el = ufl.FiniteElement("CG", mesh.ufl_cell(), 2)
+lagr_space = fem.FunctionSpace(mesh, lagr_el)
+r = fem.Function(lagr_space)
+r.interpolate(radial_distance)
```

Comment:
Instead of using a Lagrange function here, we can use
```python
x = ufl.SpatialCoordinate(mesh)
r = radial_distance(x)
```
if we use `ufl.sqrt` instead of `numpy.sqrt` in the `def radial_distance`.
This is more accurate as `r` will then be evaluated at the quadrature points.